java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -122,15 +118,62 @@ public void processElement(
             throws Exception {
         final long rowTime = ((Instant) row.getFieldAs(rowTimeFieldName)).toEpochMilli();
 
-        if (skipSameWindowOutput) {
-            // Only register timer on the event time of the row and on the row expire time.
-            ctx.timerService().registerEventTimeTimer(rowTime);
-            for (AggregationFieldsDescriptor.AggregationFieldDescriptor aggFieldDescriptor :
-                    aggregationFieldsDescriptor.getAggFieldDescriptors()) {
-                ctx.timerService()
-                        .registerEventTimeTimer(rowTime + aggFieldDescriptor.windowSizeMs);
+        long maxTriggeredTimer =
+                state.maxTriggeredTimer.value() == null
+                        ? Long.MIN_VALUE
+                        : state.maxTriggeredTimer.value();
+
+        if (rowTime + aggregationFieldsDescriptor.getMaxWindowSizeMs()
+                <= ctx.timerService().currentWatermark()) {
+            // TODO: Add metrics to record such data.
+            return;
+        }
+
+        if (rowTime > maxTriggeredTimer) {
+            state.addRow(rowTime, row, null);
+
+            if (skipSameWindowOutput) {
+                ctx.timerService().registerEventTimeTimer(rowTime);
+
+                for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
+                        aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+                    ctx.timerService().registerEventTimeTimer(rowTime + descriptor.windowSizeMs);
+                }
             }
         } else {
+            Row accumulatorStates = state.getAccumulatorStates();
+            Set<Integer> affectedFieldIndices = new HashSet<>();
+
+            for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
+                    aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+                final int aggFieldIdx = aggregationFieldsDescriptor.getAggFieldIdx(descriptor);
+                final long timeToRetract = rowTime + descriptor.windowSizeMs;
+
+                if (timeToRetract <= maxTriggeredTimer) {
+                    continue;
+                }
+
+                affectedFieldIndices.add(aggFieldIdx);
+
+                Object accumulatorState = accumulatorStates.getField(aggFieldIdx);
+                Object preAggResult = row.getField(descriptor.fieldName);
+                descriptor.aggFunc.merge(accumulatorState, preAggResult);
+                accumulatorStates.setField(aggFieldIdx, accumulatorState);
+
+                if (skipSameWindowOutput) {
+                    ctx.timerService().registerEventTimeTimer(timeToRetract);
+                }
+            }
+
+            state.addRow(rowTime, row, affectedFieldIndices);
+            state.updateAccumulatorStates(accumulatorStates);
+
+            if (skipSameWindowOutput) {
+                ctx.timerService().registerEventTimeTimer(maxTriggeredTimer + stepSizeMs);
```

Comment:
Why do we need this?

Reply:
We need to make sure that after handling late data, processfunction can output the updated agg result right in the first time, instead of waiting for other potential non-late data to trigger the output.